### PR TITLE
Add docker-build workflow (#942)

### LINKED ISSUE #942: Please add Multi-CPU architecture for Docker image
Hello, I would kindly request adding support for Multi-CPU architecture on Docker Image with building and pushing images for different architectures during CI/CD workload.

### PR DESCRIPTION
**Pull request by janw**
## Checklist
- [x] All new and existing **tests are passing**
- [x] (If docs changes needed:) I have updated the **documentation** accordingly.
- [x] I have added an entry to `CHANGES.rst` because this is a user-facing change or an important bugfix
- [x] I have written **proper commit message(s)**
      

## What changes does this Pull Request introduce?

This PR adds an automated docker build using GH actions.

Images are build automatically for the following patterns:

- PR is opened/updated: builds and tags the image but does not push it to the registry (this is done to verify a successful build; [example build](https://github.com/janw/isso/actions/runs/6102764489))
- Push to `master` (i.e. merge a PR): build and tag image with `:latest` ([example build](https://github.com/janw/isso/actions/runs/6102803470))
- Push a new tag (i.e. release new version): build and tag with
  - major version only (i.e. `:0`), and major+minor version (i.e. `:0.14`) ([example build](https://github.com/janw/isso/actions/runs/6102878038/job/16562257652))
  - full version (i.e. `:0.14.1-post.0`) ([example build](https://github.com/janw/isso/actions/runs/6102861527))

Furthermore this includes a multi-arch docker build: the image will be build for linux/amd64, and linux/arm64/v8, arguably the most common platforms for both large production environments and small-scale self-hosted applications.

Example image are available via my fork: https://github.com/janw/isso/pkgs/container/isso

## Why is this necessary?

- Removes necessity to manually build container images, and resolves #942 in the process.

### CODE REVIEW

**Comment by jelmer** [comment]
We probably need to work out where exactly we want the docker images to live, GitHub doesn't make it very easy to publish docker images cross-repo.

@ix5 do we want to move the main repo to the isso-comments org at some point

**Comment by BBaoVanC** [comment]
What other repo is there, why not just publish the images to here?

**Comment by jelmer** [comment]
This PR uses the ghcr.io/isso-comments namespace, and I think the plan is also to move this repo to GitHub.com/isso-comments/isso eventually

**Comment by ix5** [comment]
I talked to @posativ about moving the repo over to the org, but can't quite remember what prevented us from doing that.

PR itself LGTM.

File: .github/workflows/docker-build.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          flavor: |
+            latest=false
+          images: ${{ env.build_image }}
+          tags: |
+            type=ref,event=pr
+            type=semver,pattern={{major}}
```
Comment (ix5) [inline .github/workflows/docker-build.yaml:33]: Where are these values (`{{major}}`, `{{version}}`) pulled from? `git` tags?

Comment (janw) [reply .github/workflows/docker-build.yaml:33]: That's correct, it's based on semver-formatted tags. More details here: https://github.com/docker/metadata-action#typesemver

**Comment by ix5** [comment]
Just to make it clear, this is blocked until we have a decision about moving the repo over to the org.

**Comment by jelmer** [comment]
> Just to make it clear, this is blocked until we have a decision about moving the repo over to the org.

could we just publish to posativ/isso in the mean time?

**Comment by ix5** [comment]
> > Just to make it clear, this is blocked until we have a decision about moving the repo over to the org.
> 
> could we just publish to posativ/isso in the mean time?

That would be fine by me, but I don't have access to the `Settings` repo tab (for `posativ/isso`) to add a deploy key as a secret.

I _do_ have access to the [isso-comments](https://github.com/isso-comments) org and could add the deploy key as an accepted one, just as I have done for the GitHub pages repo (see [sso-comments/isso-comments.github.io](https://github.com/isso-comments/isso-comments.github.io)).

So either way, we're blocked until Martin finds time to act on this.

**Comment by janw** [comment]
Just checking in, is there anything you need from me to help move this forward? I'd be happy to assist in any way necessary.

**Comment by jelmer** [comment]
Thanks; I think we're basically just waiting from a response from @posativ

**Comment by jelmer** [comment]
The repo has now been migrated, and references are being updated in #996 so let's land this!

File: .github/workflows/docker-build.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            type=semver,pattern={{major}}.{{minor}}
+            type=semver,pattern={{version}}
+            type=raw,value=latest,enable={{is_default_branch}}
```
Comment (ix5) [inline .github/workflows/docker-build.yaml:36]: The way I read this, the action rebuilds and pushes the `:latest` image on every push to the default branch (=`master`).

**Review by ix5 — COMMENTED** [review]
I hadn't realized this at the time, but this change means we're re-building the `:latest` docker tag on _every_ push to `master`.

Up until this change, I had handled the docker image policy as `:latest` for the latest tagged release (`0.13` at the time IIRC) and `:dev` for images from the `master` branch.

Could we please change back to `:latest` being the latest *release*, not "bleeding edge" `master` branch?

**Comment by jelmer** [comment]
I think it's pretty common in docker that latest just refers to whatever the latest thing that was built and pushed was; the way the latest tag works in docker this is pretty common hard to avoid.

I think that if we want to have a docker tag for the latest release, we should just create a new ":release" tag and update that accordingly.

That said, I think it's good practice for people to specify an explicit version (v1.0) as a tag so that their service don't get silently upgraded whenever they restart their services, including any regressions that may cause.

**Pull request merged**